### DISCUSSION TITLE
문의 드려도 될까요?

### DISCUSSION THREAD

**Discussion by smartkuk** [elbakramer/koapy#2] [Q&A]
안녕하세요~

훌륭한 소스 공부 차원에서 구경(?) 하고 있습니다~
보면서 제가 못 찾은건지는 모르겠지만

조건식 로딩을 하고 KiwoomOpenApiServiceClientStubWrapper.py 소스에 조건식 이름으로 검색을 하는 부분에 함수들중에는
OnReceiveRealCondition 이벤트를 처리하는 내용이 없던데, 혹시 특별한 이유가 있나요?

**Comment by elbakramer** [comment]
안녕하세요.

우선 질문주신 내용에서 "이벤트를 처리하는 내용이 있어야 할 것 같은데 없는 부분" 을
좀 더 명확히 해주시면 더욱 정확하게 답변해드릴 수 있을 것 같습니다.

제가 유추하기로는 `GetCodeListByCondition()` 및 `GetCodeListByConditionAsStream()` 함수를 먼저 보시다가
결국에는 `self.ConditionCall()` 을 따라서 아래 부분을 보셨을 듯 한데요.
https://github.com/elbakramer/koapy/blob/6266227d51ed9aa018f9d56534b7e1002df45a86/koapy/grpc/KiwoomOpenApiServiceClientStubWrapper.py#L123-L133

여기는 사실 gRPC 서버-클라이언트 구조에서 클라이언트 측의 호출부를 좀 더 쉽게 만든 내용밖에 없습니다.
현재 구현상 클라이언트 측에서는 직접 이벤트를 처리하지는 않고 서버측에서 이벤트를 처리해서 클라이언트로 넘겨주는 형태구요.
실제 위에서 마지막 줄을 통해 클라이언트에서 서버로 요청을 하게 되고, 이후 서버측이 받아서 처리하는 부분이 아래입니다.

https://github.com/elbakramer/koapy/blob/6266227d51ed9aa018f9d56534b7e1002df45a86/koapy/grpc/KiwoomOpenApiServiceServicer.py#L134-L137

여기 함수에서 `request` 인자가 위에서 생성했던 `ConditionRequest` 메시지입니다.
함수 내부에서는 단순하게 `KiwoomOpenApiConditionEventHandler` 객체를 생성해 요청을 대신 처리하도록 되어있고요.
해당 객체를 따라가시면 말씀주신 `OnReceiveRealCondition` 이벤트를 처리하는 내용이 아래처럼 있습니다.

https://github.com/elbakramer/koapy/blob/6266227d51ed9aa018f9d56534b7e1002df45a86/koapy/grpc/event/KiwoomOpenApiEventHandlers.py#L838-L852

위의 이벤트 처리 콜백 함수 내용 외에, 서버측 `ConditionCall()` 함수 내부의 `with`/`for` 구조의 동작과 관련해서는
해당 객체의 `on_enter()` 함수 내용 및 객체가 상속하는 상위 클래스들의 구현을 참고하시면 이해하시는데 도움이 될 듯 싶습니다.

https://github.com/elbakramer/koapy/blob/6266227d51ed9aa018f9d56534b7e1002df45a86/koapy/grpc/event/KiwoomOpenApiEventHandlers.py#L793-L804

https://github.com/elbakramer/koapy/blob/6266227d51ed9aa018f9d56534b7e1002df45a86/koapy/grpc/event/KiwoomOpenApiEventHandler.py#L76-L95

주신 질문에 대한 1차 답변은 요정도로 하겠습니다.

**Comment by smartkuk** [comment]
안녕하세요 상세한 답변 주셔서 감사합니다~
제가 관련 소스 부분을 찾지 못한거네요.

키움에 TR 요청 이후에 비동기적인 이벤트 발생으로 요청의 응답을 받는 구조를 gRPC 서버 및 클라이언트 사용해서
분리해 놓으신게 더 합리적으로 보여서 저에겐 좋은 공부가 되었습니다.

제가 만든 키움모듈을 돌리면서 실시간 데이터 구독을 점차 많은 종목을 하게 되니 이벤트 수신에 부하가 슬슬 나와서
늦게 수신이 되거나 처리가 늦는거 같아서 해결 방법을 좀 찾아보고 있는데

koapy를 도입해버릴지 개선을 좀 더 해볼지 고민이 많이 되네요 ㅋㅋ
